### PR TITLE
Optionalize MappedRead.mdTag

### LINKED ISSUE #297: Remove requirement on MDTag for MappedRead
We may want to analysis mapped reads without the reference information, so we not care if they are MDTagged - should we remove this requirement? cc @ryan-williams @timodonnell

### PR DESCRIPTION
**Pull request by danvk**
Fixes #297
Supersedes #350 

This is the outcome of @timodonnell @ryan-williams and my discussion this morning. We're keeping the MD-tagged/not-tagged status of reads out of the type system for now but we'll revisit this decision in the future if it proves problematic. Attempting to access reference data on a `MappedRead` without MD tags will `throw`.

There's a new `InputFilter`, `hasMdTag`, which I've updated the existing callers and tests use. This preserves the current behavior of filtering out non-MDTag reads.

[Review on Reviewable](https://reviewable.io/reviews/hammerlab/guacamole/351)

### CODE REVIEW

File: src/main/scala/org/hammerlab/guacamole/reads/MappedRead.scala [hunk 1]
```diff
@@ -42,7 +42,7 @@ case class MappedRead(
     alignmentQuality: Int,
     start: Long,
     cigar: Cigar,
-    mdTagString: String,
+    mdTagString: Option[String],
```
Comment (ryan-williams) [inline src/main/scala/org/hammerlab/guacamole/reads/MappedRead.scala]: Orthogonal change I was gong to suggest on the other PR: I'd rather `MappedRead` take an `Option[MdTag]`, and have an alternate ctor that allows passing an `Option[String]` and making an `Option[MdTag]` out of it

File: src/main/scala/org/hammerlab/guacamole/reads/MappedRead.scala [hunk 2]
```diff
@@ -112,3 +116,6 @@ case class MissingMDTagException(record: SAMRecord)
 
 case class CigarMDTagMismatchException(cigar: Cigar, mdTag: MdTag, cause: IllegalStateException)
   extends Exception("Cigar %s seems inconsistent with MD tag %s".format(cigar.toString, mdTag.toString), cause)
+
+case class ReferenceWithoutMDTagException(read: MappedRead)
```
Comment (ryan-williams) [inline src/main/scala/org/hammerlab/guacamole/reads/MappedRead.scala]: maybe cleaner to use an interpolated string?

```
Exception(s"Attempted to get reference data for a read without an MD tag: $read")
```

File: src/main/scala/org/hammerlab/guacamole/reads/Read.scala [hunk 3]
```diff
@@ -75,6 +75,9 @@ trait Read {
   /** Whether read is from a paired-end library */
   val isPaired: Boolean
 
+  /** Whether the read has associated reference data */
+  val hasMdTag: Boolean
```
Comment (ryan-williams) [inline src/main/scala/org/hammerlab/guacamole/reads/Read.scala]: all the abstract `val`s in this `trait` don't make me feel great. Is there rhyme or reason to them? @arahuja @timodonnell

cf. http://stackoverflow.com/a/19642301/544236

**Comment by danvk** [comment]
Review status: 0 of 16 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

<sup>**[src/main/scala/org/hammerlab/guacamole/reads/MappedRead.scala, line 45 \[r1\]](https://reviewable.io:443/reviews/hammerlab/guacamole/351#-JvoQi1gPApoMheKR3aF-r1-45)** ([raw file](https://github.com/hammerlab/guacamole/blob/6907a39cba29c7f7da566f48c12c3bdaa625d1df/src/main/scala/org/hammerlab/guacamole/reads/MappedRead.scala#L45)):</sup>
Done. I wound up using a companion object & `DummyImplicit` to avoid duplicate signatures.

---

<sup>**[src/main/scala/org/hammerlab/guacamole/reads/MappedRead.scala, line 120 \[r1\]](https://reviewable.io:443/reviews/hammerlab/guacamole/351#-JvoQi1gPApoMheKR3aF-r1-120)** ([raw file](https://github.com/hammerlab/guacamole/blob/6907a39cba29c7f7da566f48c12c3bdaa625d1df/src/main/scala/org/hammerlab/guacamole/reads/MappedRead.scala#L120)):</sup>
Done.

---

<sup>**[src/main/scala/org/hammerlab/guacamole/reads/Read.scala, line 79 \[r1\]](https://reviewable.io:443/reviews/hammerlab/guacamole/351#-JvoQi1gPApoMheKR3aI-r1-79)** ([raw file](https://github.com/hammerlab/guacamole/blob/6907a39cba29c7f7da566f48c12c3bdaa625d1df/src/main/scala/org/hammerlab/guacamole/reads/Read.scala#L79)):</sup>
I changed them all to `def`.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/guacamole/351)

**Comment by ryan-williams** [comment]
lgtm

---

Review status: 0 of 16 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

<sup>**[src/main/scala/org/hammerlab/guacamole/reads/Read.scala, line 79 \[r1\]](https://reviewable.io:443/reviews/hammerlab/guacamole/351#-JvoQi1gPApoMheKR3aI-r1-79)** ([raw file](https://github.com/hammerlab/guacamole/blob/6907a39cba29c7f7da566f48c12c3bdaa625d1df/src/main/scala/org/hammerlab/guacamole/reads/Read.scala#L79)):</sup>
weird that that change is not showing up here? you mean you did that but haven't pushed it to this branch yet?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/guacamole/351)

**Comment by danvk** [comment]
Review status: 0 of 16 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

<sup>**[src/main/scala/org/hammerlab/guacamole/reads/Read.scala, line 79 \[r1\]](https://reviewable.io:443/reviews/hammerlab/guacamole/351#-JvoQi1gPApoMheKR3aI-r1-79)** ([raw file](https://github.com/hammerlab/guacamole/blob/6907a39cba29c7f7da566f48c12c3bdaa625d1df/src/main/scala/org/hammerlab/guacamole/reads/Read.scala#L79)):</sup>
Shoot, I accidentally reverted this change. Thanks for catching it. Re-done.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/guacamole/351)

**Pull request merged**